### DISCUSSION TITLE
How are you counting LL-HLS viewers?

### DISCUSSION THREAD

**Discussion by bchah** [OvenMediaLabs/OvenMediaEngine#1505] [Help]
The OvenMediaEngine API does not count the actual number of live viewers, rather the number of *connections* to the server. With LL-HLS you can reload a stream 100 times and OME will report 100 connections until your browser is closed. Sometimes connecting an LL-HLS stream will show 2 new connections right away. Or during playback another connection will sometimes be registered, if there is a network issue. Essentially, there is no way to get an accurate number of live viewers from the OME API if you are using LL-HLS playback.

See issues #1304 and #1313 for background. I have since implemented a database polling solution in my app but it's not ideal and adds a lot of overhead to the application.

My question to the community is: what are you doing to work around this issue? How are you showing the number of live viewers in your app?

**Comment by naanlizard** [comment]
We're using websockets iirc to count players, and separately we can track the number of people in the chat for a given stream, but yeah it's not great.

Would love a better way to handle it but I don't know how. Maybe something with authentications for secure playback? We're working on secure playback and the nginx proxy in front of OME is making it tricky. @hernanrz could comment more authoritatively.

**Comment by bchah** [comment]
Thanks @naanlizard - I feel less crazy for developing the workaround but it's a good excuse to break ground on a javascript based back-end instead of my aging PHP one. I use SignedPolicy so the plan is:

- open the socket
- Require the first message to include the playback URL as an input along with a random session key generated at page load
- check the signature validity by parsing and signing the policy, then comparing the value against the provided one
- if valid, store the 'session' under an in-memory object like sessions[stream_name] with values like { session_key: "x", start_time: 1706303604, data: {some: "data"} }
-  Every time a new session is registered or a socket is closed by a viewer closing the player/window, send a message to the sockets on that "channel" informing them of the new viewer count.

I'm going to try and do this as a standalone piece so if I can spin it off into something useful I will definitely start a repo.

**Reply by bchah** [reply]
Here's a skeleton of what I came up with as a NodeJS widget. You could authenticate requests to this or bake in an auth requirement to the first Websocket message. Hope it inspires a bit!

```
const WebSocket = require('ws');
const express = require('express');
const app = express();
const port = 3001;

// In-memory storage for sessions
let sessions = {};

// WebSocket server setup
const comms_session = new WebSocket.Server({ port: port });

comms_session.on('connection', (ws) => {
    let verified = false;
    ws.on('message', (message) => {



        if (!verified) {

            // perform some authentication based on the contents of the first message

        } else {

            try {
                const data = JSON.parse(message);
                const { channel_id, session_id } = data;

                if (!sessions[channel_id]) {
                    sessions[channel_id] = new Set();
                }

                sessions[channel_id].add(session_id);

                comms_session.clients.forEach((client) => {
                    if (client.readyState === WebSocket.OPEN && client !== ws) {
                        client.send(JSON.stringify({
                            "subject": "viewer_count",
                            "count": sessions[channel_id].size
                        }));
                    }
                });

                ws.on('close', () => {

                    let count = 0;

                    if (sessions[channel_id] && sessions[channel_id].has(session_id)) {
                        sessions[channel_id].delete(session_id);

                        count = sessions[channel_id].size;

                        comms_session.clients.forEach((client) => {
                            if (client.readyState === WebSocket.OPEN && client !== ws) {
                                client.send(JSON.stringify({
                                    "subject": "viewer_count",
                                    "count": count
                                }));
                            }
                        });

                    }
                });

            } catch (e) {
                console.error('Error parsing message:', e);
            }
        }
    });
});

// REST endpoint
app.get('/get_sessions', (req, res) => {
    const channel_id = req.query.channel_id;
    if (!channel_id || !sessions[channel_id]) {
        res.send("0");
    } else {
        res.send(String((sessions[channel_id] ? sessions[channel_id].size : 0)));
    }

});

app.listen(3000, () => {

});

```
